sycl/doc/extensions/proposed/sycl_ext_oneapi_forward_progress.asciidoc

```diff
@@ -0,0 +1,633 @@
+may try to acquire the lock, then work-items require stronger forward progress
+guarantees at root-group coordination scope; if each work-group has its own
+private data structure(s), and only work-items within the same work-group may
+try to acquire the same lock, then work-items require stronger forward progress
+guarantees only at work-group coordination scope.
+
+For performance reasons, a coordination scope will typically be the narrowest
+scope containing all threads that may coordinate with one another.
+====
+
+=== Forward progress guarantees
+
+The `forward_progress_guarantee` `enum` is used to represent the three classes
+of forward progress guarantee defined in the {cpp} core language specification.
+
+[source,c++]
+----
+namespace ext::oneapi::experimental::sycl {
+
+enum class forward_progress_guarantee {
+  concurrent,
+  parallel,
+  weakly_parallel
+};
+
+}
+----
+
+[NOTE]
+====
+This `enum` is defined similarly to the one defined in
+https://wg21.link/p2300[P2300: `std::execution`].
+====
+
+
+=== Execution scopes
+
+The `execution_scope` `enum` is used to represent the hierarchy of threads of
+execution within a SYCL implementation.
+
+[source,c++]
+----
+namespace ext::oneapi::experimental::sycl {
+
+enum class execution_scope {
+  work_item,
+  sub_group,
+  work_group,
+  root_group,
+};
+
+}
+----
+
+[NOTE]
+====
+The existing `memory_scope` `enum` is not used here to provide greater
+flexibility in modifying the forward progress extension in future.
+`memory_scope` is lacking values required by this extension (i.e. `root_group`)
+and has additional values that are not required here (e.g. `system`).
+====
+
+
+=== Device queries
+
+New device queries are introduced to allow developers to reason about the
+progress guarantees that can be provided by each device.
+
+The results of these queries represent the forward progress guarantees that a
+device can satisfy for each execution scope, and therefore the forward progress
+guarantees that can be requested by a kernel's <<properties,
+properties>>. These queries (and their results) must be interpreted relative to
+a coordination scope, as described previously.
+
+[NOTE]
+====
+Note that these queries cannot be used to reason about the thread of execution
+associated with the root-group, because it is currently the top of the
+hierarchy and it is therefore not possible to specify a broader coordination
+scope.
+====
+
+[NOTE]
+====
+Explicitly specifying the coordination scope ensures that the interpretation of
+forward progress queries is fixed and independent of the number of execution
+scopes in the hierarchy. This extension therefore does not prevent SYCL or
+other extensions from introducing new execution scopes.
+====
+
+
+[%header,cols="1,5,5"]
+|===
+|Device Descriptor
+|Return Type
+|Description
+
+|`template <execution_scope CoordinationScope>
+ info::device::ext_oneapi_work_group_progress_capabilities`
+|`std::vector<forward_progress_guarantee>`
```

Comment:
When I was revisiting the property descriptions on Friday, I realized that I was probably too quick to dismiss @AerialMantis' comments [here](https://github.com/intel/llvm/pull/7598#discussion_r1270832276) and [here](https://github.com/intel/llvm/pull/7598#discussion_r1270858171), and that there were some inconsistencies in the original approach.

Returning the strongest possible guarantee is sufficient for now, because we can order the existing forward progress guarantees (i.e. `concurrent` > `parallel` > `weakly_parallel`).  But we don't have any guarantees that will always be true -- if C++ (or SYCL) decides to add a new progress guarantee, then returning one value won't be enough.

The other thing is that although the concepts represented by the `enum` have an order, the values themselves aren't comparable.  So, checking if the return value is "strong enough" requires writing something like a case statement or complex if condition.  Such a condition could then break unexpectedly if a new progress guarantee is introduced.  With a vector, a developer can just call `find` to check if the vector contains the necessary guarantee, and that will always work.

I added `capabilities` to the name of these queries to align with the `memory_order` and `memory_scope` queries (which also return vectors).

Reply:
I see your point, and I don't have strong opinions here.  However, I think this extension has other assumptions about an ordered relationship among the forward progress guarantees:

* However, it remains possible for requests with different execution scopes to have overlapping coordination scopes ... In such a case, an implementation must satisfy the **strongest** request(s).

* The forward progress guarantee of a thread of execution at scope `S`, relative to coordination scope `C`, is the **weakest** of the forward progress guarantees provided by threads of execution at all scopes between `S` and `C` inclusive.

If we do end up adding new forward progress guarantees that are not ordered, we will need to consider how to reword those sections of the spec.

Reply:
That's a very good point.  Is the wording okay for now, though?

If things change, we can update the wording (e.g. to say certain combinations may be ambiguous and may throw errors, or to define a new algorithm for managing coordination scope) without changing the API.

Reply:
Yes, the wording is OK for.  I agree that we can change things later if it is necessary.